File: rd-framework/src/main/kotlin/com/jetbrains/rider/framework/base/ISerializersOwner.kt
package com.jetbrains.rider.framework.base

import com.jetbrains.rider.framework.ISerializers
import com.jetbrains.rider.framework.Protocol
import com.jetbrains.rider.util.trace

interface ISerializersOwner {
    fun register(serializers : ISerializers) {
        val key = this::class
        if (!serializers.toplevels.add(key)) return

        Protocol.initializationLogger.trace { "REGISTER serializers for ${key.simpleName}" }
        registerSerializersCore(serializers)
    }

    fun registerSerializersCore(serializers : ISerializers)
}

File: rd-cpp/src/rd_framework_cpp/src/main/base/ISerializersOwner.cpp
#include "ISerializersOwner.h"

namespace rd
{
void ISerializersOwner::registry(Serializers const& serializers) const
{
	auto it = used.insert(&serializers);
	if (!it.second)
	{
		return;
	}

	registerSerializersCore(serializers);
}
}	 // namespace rd
